### PR TITLE
Add calico as cni

### PR DESCRIPTION
**Pull request by manuelbuil**
#### Proposed Changes ####

This PR adds the required calico pieces to be able to deploy it through --cni calico

#### Types of Changes ####

New feature

#### Verification ####

Use --cni calico

#### Linked Issues ####

rancher/rke2#860
https://github.com/rancher/image-mirror/pull/99

#### Further Comments ####

Signed-off-by: Manuel Buil <mbuil@suse.com>

### CODE REVIEW

File: pkg/cli/cmds/server.go [hunk 1]
```diff
@@ -16,7 +16,7 @@ const (
 
 var (
 	DisableItems = []string{"rke2-coredns", "rke2-ingress-nginx", "rke2-kube-proxy", "rke2-metrics-server"}
-	CNIItems     = []string{"canal", "cilium"}
+	CNIItems     = []string{"canal", "cilium", "calico"}
```
[the remark this answers is not in the document]
Comment (mrostecki) [reply pkg/cli/cmds/server.go:19]: Or maybe... if you want to make it more generic

```suggestion
	// CNIItems is a mapping of a CNI plugins with a boolean value indicating
	// whether the given CNI comes with an additional CRD chart
	CNIItems = map[string]bool{"canal": false, "cilium": false, "calico": true}
```

File: pkg/cli/cmds/server.go [hunk 2]
```diff
@@ -150,6 +150,7 @@ func validateCNI(clx *cli.Context) {
 		for _, d := range CNIItems {
 			if cni != d {
 				clx.Set("disable", "rke2-"+d)
+				clx.Set("disable", "rke2-"+d+"-crd")
 			}
 		}
```
[the remark this answers is not in the document]
Comment (mrostecki) [reply pkg/cli/cmds/server.go:155]: According to my 2nd idea:

```suggestion
		for d, hasCRDChart := range CNIItems {
			if cni != d {
				clx.Set("disable", "rke2-"+d)
				if hasCRDChart {
					clx.Set("disable", "rke2-"+d+"-crd")
				}
			}
		}
```

**Review by mrostecki — CHANGES_REQUESTED** [review]
I have an another, probably more generic idea, which will not require writing down all the names

File: pkg/cli/cmds/server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (manuelbuil) [reply pkg/cli/cmds/server.go:155]: Problem of using a map is that it breaks line 26 
`Usage:  "(networking) CNI Plugin to deploy, one of none, " + strings.Join(CNIItems, ", "),``

File: pkg/cli/cmds/server.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (manuelbuil) [reply pkg/cli/cmds/server.go:19]: As we already discussed, using a map breaks line 26 which is expecting a string and it is not easy to solve:

`Usage: "(networking) CNI Plugin to deploy, one of none, " + strings.Join(CNIItems, ", "),`

**Comment by brandond** [comment]
I would prefer to keep it simple; I think the current approach of also disabling rke2-CNI-crd when rke2-CNI is disabled is fine. It hurts nothing to disable things that don't exist.

**Comment by manuelbuil** [comment]
> I would prefer to keep it simple; I think the current approach of also disabling rke2-CNI-crd when rke2-CNI is disabled is fine. It hurts nothing to disable things that don't exist.

But that would install the calico crds, or?

**Review by dweomer — APPROVED** [review]

**Comment by brandond** [comment]
> But that would install the calico crds, or?

All of the charts are enabled by default. Enabling a CNI disables both the base chart, and the CRD chart, for all other CNIs. If the other chart doesn't have a CRD chart, thats fine - the disables list is actually just a list of file manifest basenames to remove if found; it's fine to disable things that don't exist.

File: pkg/cli/cmds/server.go (hunk 1 above)
Comment (brandond) [inline pkg/cli/cmds/server.go:19]: Lets keep this sorted.
```suggestion
	CNIItems     = []string{"calico", "canal", "cilium"}
```

**Review by brandond — CHANGES_REQUESTED** [review]

File: pkg/cli/cmds/server.go (hunk 1 above)
Comment (manuelbuil) [reply pkg/cli/cmds/server.go]: done

**Review by brandond — APPROVED** [review]

File: pkg/cli/cmds/server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mrostecki) [reply pkg/cli/cmds/server.go:155]: Yeah, unfortunately seems like in Golang you have to join keys in a loop

**Pull request merged**